pkg/funder/funder.go

```diff
@@ -6,21 +6,190 @@ package funder
+
+	ctx := context.Background()
+
 	kubeClient, err := kube.NewKube()
 	if err != nil {
 		return fmt.Errorf("connecting kube client with error: %w", err)
 	}
 
-	_, err = kube.FetchNamespaceNodeInfo(context.Background(), kubeClient, cfg.Namespace)
+	key, err := wallet.GetKey()
+	if err != nil {
+		return fmt.Errorf("failed getting wallet key: %w", err)
+	}
+
+	log.Printf("using wallet address: %s", key)
+
+	ethClient, err := makeEthClient(ctx, cfg.ChainNodeEndpoint)
+	if err != nil {
+		return fmt.Errorf("failed make eth client: %w", err)
+	}
+
+	fundingWallet := wallet.New(ethClient, key)
+
+	namespace, err := kube.FetchNamespaceNodeInfo(ctx, kubeClient, cfg.Namespace)
 	if err != nil {
 		return fmt.Errorf("get node info failed with error: %w", err)
 	}
 
+	fundNodeRespC := make(chan fundNodeResp, len(namespace.Nodes))
+	for _, n := range namespace.Nodes {
+		go fundNode(ctx, fundingWallet, cfg.MinAmounts, n, fundNodeRespC)
+	}
+
+	for i := 0; i < len(namespace.Nodes); i++ {
+		resp := <-fundNodeRespC
+		if resp.err != nil {
+			log.Printf("failed to fund node (%s): %s", resp.node.Name, resp.err)
+		} else {
+			log.Printf("node funded (%s)", resp.node.Name)
+		}
+	}
+
 	return nil
 }
+
+func makeEthClient(ctx context.Context, endpoint string) (*ethclient.Client, error) {
+	rpcClient, err := rpc.DialContext(ctx, endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	return ethclient.NewClient(rpcClient), nil
+}
+
+type fundNodeResp struct {
+	node kube.Node
+	err  error
+}
+
+var (
+	ErrFailedFundingNode               = errors.New("failed funding node")
+	ErrFailedFudningNodeWithSwarmToken = errors.New("failed funding node with swarm token")
+	ErrFailedFudningNodeWithNativeCoin = errors.New("failed funding node with native coin")
+)
+
+func fundNode(
+	ctx context.Context,
+	fundingWallet *wallet.Wallet,
+	minAmounts MinAmounts,
+	node kube.Node,
+	fundNodeRespC chan<- fundNodeResp,
+) {
+	fundRespC := make(chan error, 2)
+
+	go fundNodeNativeCoin(ctx, fundingWallet, minAmounts, node, fundRespC)
+	go fundNodeSwarmToken(ctx, fundingWallet, minAmounts, node, fundRespC)
+
+	var errorMsg []string
+
+	for i := 0; i < 2; i++ {
+		if err := <-fundRespC; err != nil {
+			errorMsg = append(errorMsg, err.Error())
+		}
+	}
+
+	var err error
+	if len(errorMsg) > 0 {
+		err = fmt.Errorf("%w (%s), reason: %s", ErrFailedFundingNode, node.Name, strings.Join(errorMsg, ", "))
+	}
+
+	fundNodeRespC <- fundNodeResp{
+		node: node,
+		err:  err,
+	}
+}
+
+func fundNodeNativeCoin(
+	ctx context.Context,
+	fundingWallet *wallet.Wallet,
+	minAmounts MinAmounts,
+	node kube.Node,
+	respC chan<- error,
+) {
+	respC <- func() error {
+		cid := node.WalletInfo.ChainID
+
+		token, err := wallet.NativeCoinForChain(cid)
+		if err != nil {
+			return fmt.Errorf("%w: %s", ErrFailedFudningNodeWithNativeCoin, err)
+		}
+
+		topUpAmount := CalcTopUpAmount(minAmounts.NativeCoin, node.WalletInfo.NativeCoin, token.Decimals)
+		if topUpAmount.Cmp(big.NewInt(0)) <= 0 {
```

Comment:
Not sure if I understand this. Does this mean that we are checking whether the current amount is below or above min required amount?

Reply:
My understanding is that we just top up regardless of how much bzz or eth is in per address.

Reply:
Yes, we are check current balance and only top up to the min amount (specified when starting).

Reply:
> My understanding is that we just top up regardless of how much bzz or eth is in per address.

Why should node be funded if it already has enough bzz and eth?